node/hyperbahn/rate_limiting.js

```diff
@@ -0,0 +1,187 @@
+var DEFAULT_BUCKETS = 20;
+var TOTAL_QPS = '~~TotalQPS++';
+
+function RateLimiting(options) {
+    if (!(this instanceof RateLimiting)) {
+        return new RateLimiting(options);
+    }
+    var self = this;
+
+    assert(options.serviceDispatchHandler, 'serviceDispatchHandler is required for RateLimiting');
+    self.serviceDispatchHandler = options.serviceDispatchHandler;
+    self.egressNodes = options.serviceDispatchHandler.egressNodes;
+    self.timers = self.serviceDispatchHandler.channel.timers;
+
+    self.qpsLimits = options.qpsLimits || Object.create(null);
+    self.counters = Object.create(null);
+    self.buckets = options.buckets || DEFAULT_BUCKETS;
+    assert(self.buckets > 0 && self.buckets <= 1000, 'counter buckets should between (0 1000]');
+    self.refreshDelay = 1000 / self.buckets;
+    self.defaultServiceQpsLimit = options.defaultServiceQpsLimit || DEFAULT_SERVICE_QPS_LIMIT;
+    self.totalQpsLimit = options.totalQpsLimit || DEFAULT_TOTAL_QPS_LIMIT;
+    self.buckets = options.buckets || DEFAULT_BUCKETS;
+
+    self.refresh();
+}
+
+RateLimiting.prototype.type = 'tchannel.hyperbahn.rate-limiting';
+
+RateLimiting.prototype.refresh =
+function refresh() {
+    var self = this;
+
+    // update hte sliding window
+    var services = Object.keys(self.counters);
+    for (var i = 0; i < services.length; i++) {
+        var serviceName = services[i];
+        var counter = self.counters[serviceName];
+        var next = (counter.index + 1) % self.buckets;
+        if (counter.buckets[next]) {
+            // offset the bucket being moved out
+            counter.qps -= counter.buckets[next];
+        }
+        assert(counter.qps >= 0, 'qps shouls always be larger equal to 0');
+        counter.index = next;
+        counter.buckets[counter.index] = 0;
+        if (serviceName === TOTAL_QPS) {
+            // total QPS limit never change
+            continue;
+        }
+        var limit = self.qpsLimits[serviceName] || self.defaultServiceQpsLimit;
+        counter.qpsLimit = limit / self.egressNodes.kValueFor(serviceName);
+    }
+
+    self.refreshTimer = self.timers.setTimeout(
+        function refresh() {
+            self.refresh();
+        },
+        self.refreshDelay
+    );
+};
+
+RateLimiting.prototype.removeCounter =
+function removeCounter(serviceName) {
+    var self = this;
+    delete self.counters[serviceName];
+};
+
+RateLimiting.prototype.isExitFor =
+function isExitFor(serviceName) {
+    var self = this;
+
+    var chan = self.serviceDispatchHandler.channel.subChannels[serviceName];
+    if (!chan) {
+        return self.egressNodes.isExitFor(serviceName);
+    }
+
+    return chan.serviceProxyMode === 'exit';
+};
+
+RateLimiting.prototype.createCounter =
+function createCounter(serviceName) {
+    var self = this;
+
+    var counter = Object.create(null);
+    counter.index = 0;
+    counter.qps = 0;
+    counter.buckets = [];
+    counter.buckets.length = self.buckets;
```

Comment:
uninitiazilezd / undefined new buckets? can we just initialize this with a for loop please?